File: sdks/csharp/src/Sudomimus.Token/JwtToken.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace Sudomimus.Token;

/// <summary>
/// A parsed Sudomimus JWT. Exposes the header (envelope claims), body
/// (caller-specific claims), and signing input so callers can re-verify
/// signatures with their own public key.
/// </summary>
/// <typeparam name="TBody">The body claim shape, e.g. <see cref="AccessTokenBody"/>.</typeparam>
public sealed class JwtToken<TBody>
    where TBody : class
{
    /// <summary>The raw, dot-joined JWT exactly as received on the wire.</summary>
    public string Raw { get; }

    /// <summary>
    /// The exact bytes that the signature signs:
    /// <c>headerSegment + "." + bodySegment</c>, UTF-8 encoded. Required
    /// for signature verification because <c>@sudoo/jwt</c> uses a
    /// non-standard mix of base64 (header/body) and base64url (signature)
    /// — re-encoding from the decoded header/body would not match.
    /// </summary>
    public byte[] SigningInput { get; }

    /// <summary>Raw base64url-decoded signature bytes.</summary>
    public byte[] Signature { get; }

    public JwtHeader Header { get; }
    public TBody Body { get; }

    internal JwtToken(string raw, byte[] signingInput, byte[] signature, JwtHeader header, TBody body)
    {
        Raw = raw;
        SigningInput = signingInput;
        Signature = signature;
        Header = header;
        Body = body;
    }

    /// <summary>
    /// Verify the signature with the given PEM-encoded RSA public key.
    /// Returns <c>true</c> when the signature matches.
    /// </summary>
    public bool VerifySignature(string publicKeyPem)
    {
        using var rsa = RSA.Create();
        rsa.ImportFromPem(publicKeyPem);
        return rsa.VerifyData(
            SigningInput,
            Signature,
            HashAlgorithmName.SHA256,
            RSASignaturePadding.Pkcs1);
    }

    /// <summary>
    /// Verify the token's <c>exp</c> claim against the given moment.
    /// Returns <c>true</c> when the token has not yet expired.
    /// </summary>
    public bool VerifyExpiration(DateTimeOffset now)
    {
        if (Header.ExpiresAt is null)
        {
            return false;
        }
        var expiresAt = DateTimeOffset.FromUnixTimeSeconds(Header.ExpiresAt.Value);
        return now < expiresAt;
    }
}

/// <summary>
/// Implements the on-wire JWT layout used by <c>@sudoo/jwt</c>:
/// <list type="bullet">
///   <item><description>Header segment: standard base64 of JSON, <c>=</c> padding stripped.</description></item>
///   <item><description>Body segment: standard base64 of JSON, <c>=</c> padding stripped.</description></item>
///   <item><description>Signature segment: base64url of RSA-SHA256 bytes, <c>=</c> padding stripped.</description></item>
///   <item><description>Signing input: literal <c>headerSegment.bodySegment</c> string, UTF-8.</description></item>
/// </list>
/// </summary>
internal static class JwtCodec
{
    public static string EncodeJsonSegment<T>(T value, JsonSerializerOptions? options = null)
    {
        var json = JsonSerializer.Serialize(value, options);
        var bytes = Encoding.UTF8.GetBytes(json);
        return StripPadding(Convert.ToBase64String(bytes));
    }

    public static byte[] DecodeStandardBase64Segment(string segment)
    {
        var padded = PadBase64(segment);
        return Convert.FromBase64String(padded);
    }

    public static byte[] DecodeBase64UrlSegment(string segment)
    {
        var translated = segment.Replace('-', '+').Replace('_', '/');
        return DecodeStandardBase64Segment(translated);
    }

    public static string EncodeBase64UrlSignature(byte[] signature)
    {
        var b64 = StripPadding(Convert.ToBase64String(signature));
        return b64.Replace('+', '-').Replace('/', '_');
    }

    private static string StripPadding(string b64) => b64.TrimEnd('=');

    private static string PadBase64(string b64)
    {
        var remainder = b64.Length % 4;
        return remainder == 0 ? b64 : b64 + new string('=', 4 - remainder);
    }
}

File: sdks/java/token/src/main/java/com/sudomimus/token/JwtToken.java
package com.sudomimus.token;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.Signature;
import java.security.spec.X509EncodedKeySpec;
import java.time.Instant;
import java.util.Base64;

/**
 * A parsed Sudomimus JWT. Keeps the raw on-wire signing input so signature
 * verification operates on the literal bytes that were signed — re-encoding
 * the deserialized header/body could introduce key-ordering or whitespace
 * drift.
 *
 * @param <TBody> the body claim shape, e.g. {@link AccessTokenBody}.
 */
public final class JwtToken<TBody> {

    private static final String PEM_HEADER = "-----BEGIN PUBLIC KEY-----";
    private static final String PEM_FOOTER = "-----END PUBLIC KEY-----";

    private final String raw;
    private final byte[] signingInput;
    private final byte[] signature;
    private final JwtHeader header;
    private final TBody body;

    JwtToken(String raw, byte[] signingInput, byte[] signature, JwtHeader header, TBody body) {
        this.raw = raw;
        this.signingInput = signingInput;
        this.signature = signature;
        this.header = header;
        this.body = body;
    }

    public String getRaw() { return raw; }
    public byte[] getSigningInput() { return signingInput.clone(); }
    public byte[] getSignature() { return signature.clone(); }
    public JwtHeader getHeader() { return header; }
    public TBody getBody() { return body; }

    /** Returns true when the token's {@code exp} claim is in the future relative to {@code now}. */
    public boolean verifyExpiration(Instant now) {
        if (header.expiresAt == null) {
            return false;
        }
        Instant expiresAt = Instant.ofEpochSecond(header.expiresAt);
        return now.isBefore(expiresAt);
    }

    /** Returns true when the RSA-SHA256 signature matches the given PEM-encoded public key. */
    public boolean verifySignature(String publicKeyPem) {
        try {
            PublicKey pub = parseRsaPublicKey(publicKeyPem);
            Signature sig = Signature.getInstance("SHA256withRSA");
            sig.initVerify(pub);
            sig.update(signingInput);
            return sig.verify(signature);
        } catch (Exception e) {
            return false;
        }
    }

    private static PublicKey parseRsaPublicKey(String pem) throws Exception {
        String body = pem
                .replace(PEM_HEADER, "")
                .replace(PEM_FOOTER, "")
                .replaceAll("\\s+", "");
        byte[] der = Base64.getDecoder().decode(body);
        X509EncodedKeySpec spec = new X509EncodedKeySpec(der);
        return KeyFactory.getInstance("RSA").generatePublic(spec);
    }

    static byte[] signingInputBytes(String headerSeg, String bodySeg) {
        return (headerSeg + "." + bodySeg).getBytes(StandardCharsets.UTF_8);
    }
}
